### PR TITLE
[13.0][ADD] l10n_th_tin_service

### PR DESCRIPTION
**Pull request by poonlap**
Refer to the [PR #213](https://github.com/OCA/l10n-thailand/pull/213), here is revised version.
Now, the configuration is only import Thai province names.

### CODE REVIEW

**Comment by poonlap** [comment]
@kittiu  May I ask, to pass the codecov, I should write the test unit. Is my understanding correct?

**Comment by kittiu** [comment]
Yes, and it is highly recommended to do so. Because, if you have a good test, your module can get migrated to the next version easily (and people will likely want to do).

**Comment by kittiu** [comment]
@poonlap not sure this is ready for review. But I just take a peek and still can't get it to work.

Steps:
1. New database with demo data.
2. Install l10n_th_tin_service -> success install.
3. Import from Geonames
4. Create a contact, and test with a Tax ID. On change will load the page a while, before throw error,

```
  File "/usr/local/lib/python3.6/site-packages/zeep/transports.py", line 126, in _load_remote_data
    response = self.session.get(url, timeout=self.load_timeout)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='rdws.rd.go.th', port=443): Max retries exceeded with url: /serviceRD3/checktinpinservice.asmx?wsdl (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fa3b64d9f28>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))
```

May be I miss something. Also, I have quite preliminary question.

* Will the cert key last long? How do you get it anyway? https://github.com/OCA/l10n-thailand/pull/213#discussion_r536513478
* Does the tin service support English address too?

Thank you!

@sansirit can you help review this?

**Comment by poonlap** [comment]
@kittiu  Thank you for your time.
> 
> @poonlap not sure this is ready for review. But I just take a peek and still can't get it to work.
> 
> Steps:
> 
>     1. New database with demo data.
> 
>     2. Install l10n_th_tin_service -> success install.
> 
>     3. Import from Geonames

I supposed you imported Thai's provinces. 

> 
>     4. Create a contact, and test with a Tax ID. On change will load the page a while, before throw error,
> 
> 
> ```
>   File "/usr/local/lib/python3.6/site-packages/zeep/transports.py", line 126, in _load_remote_data
>     response = self.session.get(url, timeout=self.load_timeout)
>   File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 546, in get
>     return self.request('GET', url, **kwargs)
>   File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 533, in request
>     resp = self.send(prep, **send_kwargs)
>   File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 646, in send
>     r = adapter.send(request, **kwargs)
>   File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 516, in send
>     raise ConnectionError(e, request=request)
> requests.exceptions.ConnectionError: HTTPSConnectionPool(host='rdws.rd.go.th', port=443): Max retries exceeded with url: /serviceRD3/checktinpinservice.asmx?wsdl (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fa3b64d9f28>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))
> ```
> 
> May be I miss something. Also, I have quite preliminary question.

Testing in Travis-CI was OK. 

image


It looks like the error when connecting (zeep + url3). I'm not sure is that because the cert or just unstable connection, or library version.

In the file res_partner.py, can you try  changing the code 'sess.verify = False' to bypass not using cert.


> 
>     * Will the cert key last long? How do you get it anyway? [#213 (comment)](https://github.com/OCA/l10n-thailand/pull/213#discussion_r536513478)
> 
>     * Does the tin service support English address too?

No, only Thai language information is available.

> > 
> @sansirit can you help review this?

**Comment by poonlap** [comment]
The client cert is a link at the end of this page.
https://www.rd.go.th/publish/42535.0.html

But you will get 2 certs. I just combined into one. 
Actually, the RD's server is not properly set. I'm not sure if we can distribute the cert.
Should we bypass and use non-ssl connection (sess.verify = False) by default?

**Comment by poonlap** [comment]
I try to figure out how to login to runbot, but I don't know how.

Instead, I created a docker image. You may try. 
$ docker run -d -e POSTGRES_USER=odoo -e POSTGRES_PASSWORD=odoo -e POSTGRES_DB=postgres --name db postgres:10
$ docker run -p 8069:8069 --name odoo --link db:db -t poonlap/odoo-th:13.0-tin-service

**Comment by poonlap** [comment]
Hello Khun @kittiu  and @sansirit . Are you available to review the code? 
If there is anything I can do, please let me know.

**Comment by kittiu** [comment]
Sorry for the delayed krub. Here are some reviews krub.

* Functional tested, looking good now.
* Please add zeep here, https://github.com/poonlap/l10n-thailand/blob/13.0-add-l10n_th_tin_service/requirements.txt
* You many already knew, you may squash commits into 1 commit later.

Questions:

* I see you have option sess.verify. I see that cert is marked to expired on 2565? Will we have problem after that?
* What is difference between using cert and not using cert. (will it be ok just not using cert?, what is your opinion on this)

**Comment by kittiu** [comment]
I think this is a very useful module, by the way :).

File: l10n_th_tin_service/models/__init__.py
Comment (newtratip) [inline l10n_th_tin_service/models/__init__.py:1]: Please add `# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)` on line 1 same as `l10n_th_tin_service/__init__.py` file

**Review by kittiu — APPROVED** [review]

File: l10n_th_tin_service/tests/__init__.py
Comment (newtratip) [inline l10n_th_tin_service/tests/__init__.py:1]: Please add `# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)` on line 1 same as `l10n_th_tin_service/__init__.py` file

File: l10n_th_tin_service/readme/USAGE.rst
Comment (newtratip) [inline l10n_th_tin_service/readme/USAGE.rst]:
```suggestion
.. figure:: ../static/description/geonames.png
```

File: l10n_th_tin_service/readme/USAGE.rst
Comment (newtratip) [inline l10n_th_tin_service/readme/USAGE.rst]:
```suggestion
 .. figure:: ../static/description/input_tin.png
```

File: l10n_th_tin_service/readme/USAGE.rst
Comment (newtratip) [inline l10n_th_tin_service/readme/USAGE.rst]:
```suggestion
 .. figure:: ../static/description/autofill_sample.png
```

File: l10n_th_tin_service/readme/USAGE.rst
Comment (newtratip) [inline l10n_th_tin_service/readme/USAGE.rst]:
```suggestion
 .. figure:: ../static/description/main_branch.png
```

File: l10n_th_tin_service/readme/USAGE.rst
Comment (newtratip) [inline l10n_th_tin_service/readme/USAGE.rst]:
```suggestion
 .. figure:: ../static/description/another_branch.png
```

File: l10n_th_tin_service/readme/USAGE.rst
Comment (newtratip) [inline l10n_th_tin_service/readme/USAGE.rst]:
```suggestion
 .. figure:: ../static/description/tin_validation.png
```

File: l10n_th_tin_service/readme/USAGE.rst
Comment (newtratip) [inline l10n_th_tin_service/readme/USAGE.rst]:
```suggestion
 .. figure:: ../static/description/branch_validation.png
```

File: l10n_th_tin_service/readme/USAGE.rst
Comment (newtratip) [inline l10n_th_tin_service/readme/USAGE.rst]:
```suggestion
 .. figure:: ../static/description/branch_format.png
```

File: l10n_th_tin_service/readme/USAGE.rst
Comment (newtratip) [inline l10n_th_tin_service/readme/USAGE.rst]:
```suggestion
* Go to Contacts > Configuration > Localization > Import from Geonames.
```

File: l10n_th_tin_service/views/res_partner_view.xml
Comment (newtratip) [inline l10n_th_tin_service/views/res_partner_view.xml]: inherit xml id should same with original view
```suggestion
    <record id="view_partner_form" model="ir.ui.view">
```

File: l10n_th_tin_service/views/res_partner_view.xml
Comment (newtratip) [inline l10n_th_tin_service/views/res_partner_view.xml]: name_company is in l10n_th_partner module, I think this view should inherit on that.
```suggestion
        <field name="inherit_id" ref="l10n_th_partner.view_partner_form" />
```

File: l10n_th_tin_service/views/res_partner_view.xml
Comment (newtratip) [inline l10n_th_tin_service/views/res_partner_view.xml:8]: Tax ID and Tax Branch display 2 position.
Selection_153

**My Question**
1. Tax ID and Tax Branch want to move after Company Name ?
2. For individual (not company), Can we fill in only Tax ID ?

File: l10n_th_tin_service/tests/test_tin_service.py
Comment (newtratip) [inline l10n_th_tin_service/tests/test_tin_service.py:1]: Please add 

>  Copyright 2020 Poonlap V. <poonlap@tanabutr.co.th>
>  License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)

on line 1

**Review by newtratip — COMMENTED** [review]
Little comment

**Comment by newtratip** [comment]
Can you generate README follow this https://github.com/OCA/maintainer-tools/blob/master/README.md#readme-generator

File: l10n_th_tin_service/models/res_partner.py
Comment (sansirit) [inline l10n_th_tin_service/models/res_partner.py:121]: The `street_map` is a constant. So, please consider change it to a private class variable. And, please use tuple instead of list, therefore no other method can change its value.

```suggestion
_street_map = (("vBuildingName", "vRoomNumber", "vFloorNumber", "vHouseNumber", "vStreetName", "vSoiName" ),
        ("vVillageName", "vMooNumber", "vThambol") )
```

File: l10n_th_tin_service/models/res_partner.py
Comment (sansirit) [inline l10n_th_tin_service/models/res_partner.py:139]: The branch validation should be implemented separately in a dedicated `@api.constrains('branch')` method.

```suggestion
    # put this outside of the _onchange_vat_branch() method
    @api.constrains('branch', 'country_id')
    def _check_branch(self):
        for record in self:
            if "TH" != record.country_id.code or not record.branch:
                continue
            check_branch = re.compile(r"^\d{5}$")
            match = check_branch.match(self.branch)
            if match is None:
                return {"warning": {
                            "title": _("Branch validation failed"),
                            "message": _("Branch number %s must be 5 digits." % self.branch), }
                        }            
```

File: l10n_th_tin_service/models/res_partner.py
Comment (sansirit) [inline l10n_th_tin_service/models/res_partner.py]: The module should automatically add `0` padding, i.e. `'{:05d}'.format(int(self.branch))`. This is to improve UX of the modules.

```suggestion
        if not self.branch:
            self.branch = "00000"
        else:
            self.branch = "{:05d}".format(int(self.branch))
```

File: l10n_th_tin_service/models/res_partner.py
Comment (sansirit) [inline l10n_th_tin_service/models/res_partner.py:141]: In case of an error from calling the API, the `data` is `False`.  `_logger.log(DEBUG, "False")` does not make any sense for debugging. Please put `data = _get_info_rd_vat_service(self.vat, self.branch)` in a `try...except...` block.
```suggestion
                try:
                    data = ResPartner.get_info_rd_vat_service(self.vat, self.branch)
                    _logger.log(DEBUG, pprint.pformat(data))
                except:
                    _logger.log(DEBUG, _("There is an error calling RD's API"))  # or other suitable error message
```

File: l10n_th_tin_service/models/res_partner.py
Comment (sansirit) [inline l10n_th_tin_service/models/res_partner.py]: The module should have a dedicated `@api.constrains('vat')` method for validate the TIN number if and only if the company address is in Thailand. The TIN number must be 13-digit long and can be validated using the same check digit calculation as Thailand National ID.

To prevent forging TIN, checking with RD webservice is a great idea (as your implementation here). However, forged TIN would not cause "TIN is valid. Branch is not valid". So, a more sensible error message will help improve UX greatly.

```suggestion
                if not data:
                    return {
                        "warning": {
                            "title": _("Validation failed."),
                            "message": _(
                                "TIN: %s and branch: %s is not found on the RD database." % (self.vat, self.branch)
                            ),
                        }
                    }

    # put this outside of the _onchange_vat_branch() method
    @api.constrain('vat', 'country_id')
    def _check_vat(self):
        for record in self:
            if "TH" != record.country_id.code or not record.vat:
                continue
            vat= record.vat
            if len(vat) != 13 or not _id.isdigit():
                raise UserError(_("TIN must be 13 digits"))
            x = sum([(13 - i) * int(vat[i]) for i in range(12)]) % 11
            checksum = (11 - x) % 10
            if int(vat[12]) != checksum:
                raise UserError(_("Invalid TIN"))
```
The above code is partially obtained from #222

File: l10n_th_tin_service/models/res_partner.py
Comment (sansirit) [inline l10n_th_tin_service/models/res_partner.py:128]: Please use this pattern to improve code readability.
`if not self.vat: return`
`... (remaining code of the method)`
instread of 
`if self.vat:`

> `    ... (remaining code of the method)`

` # no else statement

```suggestion
        if not self.vat:
            return
            
        # remove an indentation (4 spaces) from the remaining code 
```

File: l10n_th_tin_service/models/res_partner.py
Comment (sansirit) [inline l10n_th_tin_service/models/res_partner.py:129]: Please use this pattern to improve code readability.
`if ResPartner.check_rd_tin_service(self.vat):`
> `    ... (remaining code of the method)`

`else:`
> `    return { ... }`

to
`if not ResPartner.check_rd_tin_service(self.vat):`
> `    return { ... }`
`    `

`... (remaining code of the method)`

```suggestion
            if not ResPartner.check_rd_tin_service(self.vat):
                return {"warning": {"title": _("Validation failed."), "message": _("Failed to verify TIN or PIN %s." % self.vat), },}
                
            # remove an indentation (4 spaces) from the remaining code 
```

File: l10n_th_tin_service/models/res_partner.py
Comment (sansirit) [inline l10n_th_tin_service/models/res_partner.py]: This `else` block should be put right after `if` statement to improve code readability as explain in the suggestion on line 127
```suggestion

```

File: l10n_th_tin_service/models/res_partner.py
Comment (sansirit) [inline l10n_th_tin_service/models/res_partner.py:93]: Please use this pattern to improve code readability.

```suggestion
        if odata["vmsgerr"]:
            return False
        for key, value in odata.items():
            if (value and value["anyType"][0] != "-" and key not in {"vNID", "vBusinessFirstDate"} ):
                data[key] = value["anyType"][0]
        return data
```

File: l10n_th_tin_service/models/res_partner.py
Comment (sansirit) [inline l10n_th_tin_service/models/res_partner.py:79]: This statement calls an external API. There might be an unknown error. Therefore, please enclose it with a `try... except...` block.

```suggestion
        try:
            result = cl.service.Service(
                "anonymous",
                "anonymous",
                TIN=tin,
                ProvinceCode=0,
                BranchNumber=branch,
                AmphurCode=0,
            )
        except:
            raise(UserError(_("There is an error connecting to RD webservice. The TIN validation is skipped."))  # or other sensible error message
```

File: l10n_th_tin_service/models/res_partner.py
Comment (sansirit) [inline l10n_th_tin_service/models/res_partner.py:47]: The expressions `result = cl.service. ...` should be enclosed in a try block with a generic except, because this expression will call external API which might cause an unknown error.

```suggestion
        try:
            result = cl.service.ServiceTIN("anonymous", "anonymous", tin)
        except:
            raise(UserError(_("There is an error connecting to RD webservice. The TIN validation is skipped."))  # or other sensible error message
```

File: l10n_th_tin_service/models/res_partner.py
Comment (sansirit) [inline l10n_th_tin_service/models/res_partner.py:28]: This should be a private method.

```suggestion
    @staticmethod
    def _check_rd_tin_service(tin):
```

File: l10n_th_tin_service/models/res_partner.py
Comment (sansirit) [inline l10n_th_tin_service/models/res_partner.py:53]: This should be private methods

```suggestion
    @staticmethod
    def _get_info_rd_vat_service(tin, branch=0):
```

File: l10n_th_tin_service/models/res_partner.py
Comment (sansirit) [inline l10n_th_tin_service/models/res_partner.py:96]: It should check and process the method if and only if `self.country_id.code == 'TH'` 

```suggestion
    @api.onchange("vat", "branch", "country_id")
    def _onchange_vat_branch(self):
        if "TH" != record.country_id.code or not (record.branch and record.vat):
            return
```

File: l10n_th_tin_service/views/res_partner_view.xml
Comment (sansirit) [inline l10n_th_tin_service/views/res_partner_view.xml:13]: the `placeholder` should be updated from `e.g. BE0477472701` to something like `e.g. 0105519005906`.
```suggestion
                    <field
                        name="vat"
                        class="oe_edit_only"
                        placeholder="e.g. 0105519005906"
                        attrs="{'invisible': [('is_company', '=', False)], 'required': [('is_company', '=', True)]}"
                    />
```

File: l10n_th_tin_service/views/res_partner_view.xml
Comment (sansirit) [inline l10n_th_tin_service/views/res_partner_view.xml:21]: The existing `vat` and `branch` fields should be `invisible` or `readonly` on the `view_partner_form` .

```suggestion
            </field>
            <field name="vat" position="attributes">
                <attribute name="attrs">{'invisible': [('country_code', '!=', 'TH')]}</attribute>
            </field>
            <field name="branch" position="attributes">
                <attribute name="attrs">{'invisible': [('country_code', '!=', 'TH')]}</attribute>
            </field>
        </field>
```

**Review by sansirit — CHANGES_REQUESTED** [review]
This is a fantastic module for Thai localization.
I found somethings to improve as suggested on the code.

In addition to those suggestion. I think the module should provide i18n translation for Thai (`th_TH`) language of all translatable strings. I listed them here, but I might miss something.
- "Branch validation failed"
- "Branch number %s must be 5 digits."
- "Validation failed."
- "TIN %s is valid. Branch %s is not valid."
- "Can not get info from TIN"
- "%s is valid but no address information."
- "Failed to verify TIN or PIN %s."

THANK YOU SO MUCH for contributing and creating this great module.

**Comment by sansirit** [comment]
> Questions:
> 
> * I see you have option sess.verify. I see that cert is marked to expired on 2565? Will we have problem after that?
> * What is difference between using cert and not using cert. (will it be ok just not using cert?, what is your opinion on this)

Although the CA root certificate is not trusted, I think using cert is more preferable. The current certificate has been issued on 2/5/2555 BE and it will be expired on 5/5/2565 BE. So, they issued a 10-year certificate. Therefore, we will need to monitor for a new certificate and update the certificate file every 10 years.

To avoid getting error from certificate, we might put try to catch exception `requests.exceptions.SSLError`. However, I am not sure which exception zeep will throw if the certificate expired.

The different between using cert and not using cert are:
- if not using cert, someone might create a fake rdws server with a fake DNS to bypass all checking.
- if using cert, maintaining/updating an untrusted certificate is needed. This will cause a maintainability of an Odoo/OCA system that used this module.

**Comment by kittiu** [comment]
@poonlap do you plan to attend the comments?

**Comment by poonlap** [comment]
> 
> 
> @poonlap do you plan to attend the comments?

It has been extremely with daily work in last 2 weeks. I will follow up soon.

File: l10n_th_tin_service/views/res_partner_view.xml
Comment (poonlap) [reply l10n_th_tin_service/views/res_partner_view.xml:8]: Answer:
1. Ideally, yes. I want to move "Tax ID" and "Branch" after Company Name only when the radio button "Company" is selected. 
2. When "Individual" is selected, you will not see duplicated "Tax ID" and "Branch". They are at the original position which after the address.

**Comment by poonlap** [comment]
I tried to apply all code suggestions as possible. 
For some suggestions, I need time to approve the code, but it seems I couldn't do it now.

**Comment by kittiu** [comment]
> I tried to apply all code suggestions as possible.
> For some suggestions, I need time to approve the code, but it seems I couldn't do it now.

No worry, just do it as time allowed. Feel free to nag me, if I do not response.

**Comment by poonlap** [comment]
@kittiu Thank you for your message.
I will not touch this code further because I don't have time for coding and need to get back to my daily work. I'm sorry.

I merged all suggestions of Khun @newtratip  (view + README) and some suggestions (not all) from Khun @sansirit to this branch already.  Sorry to Khun @sansirit , your suggestions are too complicated to me and need more time to implement.  

Please consider to do anything which you think necessary.

**Comment by kittiu** [comment]
ping @sansirit

**Comment by sansirit** [comment]
The RD certificate will be expired on May 2022. RD has not release a new CA yet. I think we can merge this PR after resolve the conflicting manifest file.

**Comment by kittiu** [comment]
@poonlap can you rebase this PR

**Comment by wipawii** [comment]
> The RD certificate will be expired on May 2022. RD has not release a new CA yet. I think we can merge this PR after resolve the conflicting manifest file.

@sansirit Can you share what RD certificate is? What it is for? Thanks in advance.

**Comment by sansirit** [comment]
> @sansirit Can you share what RD certificate is? What it is for? Thanks in advance.

The RD certificate is used to get access to RD services used in this module. For more details of the RD services and the certificate file are available [here](https://www.rd.go.th/publish/42535.0.html).

**Comment by kittiu** [comment]
Let's merge this PR for now. We can improve it along the way if face any issues.

**Comment by kittiu** [comment]
> Let's merge this PR for now. We can improve it along the way if face any issues.

Sorry, I still cannot merge this PR. I found that this PR may not be correctly rabased. 
https://github.com/OCA/l10n-thailand/pull/215/files

For example, it includes the chanbe in l10n_th_base_location which is already in 13.0.

May be the best way, is to clone the new 13.0, and make the clean PR again.

**Comment by poonlap** [comment]
> 
> 
> > Let's merge this PR for now. We can improve it along the way if face any issues.
> 
> Sorry, I still cannot merge this PR. I found that this PR may not be correctly rabased.
> https://github.com/OCA/l10n-thailand/pull/215/files
> 
> For example, it includes the chanbe in l10n_th_base_location which is already in 13.0.
> 
> May be the best way, is to clone the new 13.0, and make the clean PR again.

@kittiu  I have just synced  the branch 13.0-add-l10n_th_tin_service with upstream/13.0. 
Could you check if you can merge?

**Comment by kittiu** [comment]
Sorry for delayed, just notice it.

**Pull request closed without merging**
